### PR TITLE
interface: add transitional browser-support interface

### PR DESCRIPTION
**Pull request by jdstrand**
Modern browsers use a wide variety of techniques to sandbox themselves and monitor resources. In the interest of time, this PR implements a transitional interface for browsers so that major browser vendors like Google and Mozilla can use it to snap their browsers right away and so that embedded webviews using the Chromium Content API work.

The standalone browsers from major vendors (ie, Google and Mozilla) use internal sandboxes that require either significant permissions (Google's setuid sandbox) or use user namespaces that aren't significant privileges within the kernel, but require granting significant permissions in the snappy security sandbox due to limitiations of kernel LSMs and user namespaces at this time. In an effort to allow embedded webviews relatively safe policy but also allow major vendors to continue to use their sandboxing, this PR implements the optional 'allow-browser-sandbox' attribute.

When `allow-browser-sandbox: false` (the default), it allows:
1. writes to @{PROC}/@{pid}/oom_score_adj
2. writes to /var/tmp/etilqs_*
3. reads and writes to /dev/shm/.org.chromium.Chromium.* and /dev/shm/.com.google.Chrome.*
4. setpriority without args

'1' requires updates to the LSM to properly mediate. '2' and '3' require upstream code changes or the planned preload library and '4' will be addressed with seccomp arg filtering once it is available in 16.04.

When `allow-browser-sandbox: true`, in addition to the above it also allows:
1. access to some directories the give small information leaks on installed software
2. ptrace (trace) for processes within the snap
3. anonymous socket access to processes within the snap
4. various capabilities to allow privilege dropping in from the setuid sandbox
5. setuid and setgid for privilege dropping
6. the unshare syscall for user namespaces
7. quotactl for firefox

'2' requires a kernel patch to be safe (undergoing upstream and Ubuntu review), '3' is safe and planned for another PR for the default template, '4' we might be able to whittle down when we enforce user namespaces and disallow setuid sandbox, '5' and '7' might be able to have some refinements when seccomp arg filtering is available in 16.04.

This is tested to work with:
- chrome (using `allow-browser-sandbox: true` with both setuid sandbox and --disable-setuid-sandbox)
- chromium (using `allow-browser-sandbox: true` with both setuid sandbox and --disable-setuid-sandbox)
- firefox (using `allow-browser-sandbox: true`)
- electron-quick-start (using `allow-browser-sandbox: false`)
- vscode (electron, using `allow-browser-sandbox: false`)

The idea here is that because this is a transitional interface to enable existing browsers on classic systems, we auto-connect them. The store would trigger manual review if specifying 'plugs: [browser]' with 'daemon: ...' in the yaml or when specifying `allow-browser-sandbox: true` (until assertions for this are available). We would then allow reputable browser vendors to use `allow-browser-sandbox: true` while letting developers who use electron or embed chromium content api in various ways to be unencumbered. Once sandboxing mechanisms improve and/or upstreams adjust to working within the snappy sandbox, we can deprecate this interface (and we can fine-tune it as we go).

One point worth noting is that CLONE_NEWUSER does not work properly when running under newer snap-confine, regardless of devmode. This will be investigated and this interface's policy adjusted accordingly.

### CODE REVIEW

File: docs/interfaces.md [hunk 1]
```diff
@@ -107,6 +107,17 @@ Auto-Connect: no
 
 ## Supported Interfaces - Advanced
 
+### browser
```
Comment (niemeyer) [inline docs/interfaces.md]: We another name for the interface, as saying plugs: [browser] should imply _talking_ to a browser rather than _being_ a browser, per existing conventions.

File: docs/interfaces.md [hunk 2]
```diff
@@ -107,6 +107,17 @@ Auto-Connect: no
...
+
+Can access files and IPC needed by modern browsers. This interface is
+intended to be used when using an embedded Chromium Content API or using the
+sandboxes in major browsers from vendors like Google and Mozilla.
+
+Usage: reserved
+Auto-Connect: yes
+Attributes:
+- allow-browser-sandbox: true|false (defaults to ``false``)
```
Comment (niemeyer) [inline docs/interfaces.md]: `sandbox: true` should be enough, given the context.

**Comment by jdstrand** [comment]
Downgraded to ubuntu-core-launcher in 16.04 and I was able to use CLONE_NEWUSER, so added the necessary extra files. Still need to investigate the problem with newer snap-confine.

File: docs/interfaces.md (hunk 1 above)
Comment (jdstrand) [inline docs/interfaces.md]: Regarding the name, yeah, I had a tough time with this one. It doesn't fit control, observe, bind, or manage. I'm definitely open to suggestions.

File: docs/interfaces.md (hunk 2 above)
Comment (jdstrand) [inline docs/interfaces.md]: Re `sandbox: true`, I initially did that, but then I though people might get confused and think it referred to the snappy sandbox. Happy to change it if you think that isn't an issue.

File: interfaces/builtin/unity7.go [hunk 3]
```diff
@@ -249,21 +249,21 @@ dbus (send)
 # dbusmenu
 dbus (send)
     bus=session
-    path=/MenuBar{,/[0-9]*}
+    path=/{MenuBar{,/[0-9A-F]*},com/canonical/menu/[0-9A-F]*}
```
Comment (niemeyer) [inline interfaces/builtin/unity7.go]: Is this leaking in?  Seems unrelated.

File: docs/interfaces.md (hunk 2 above)
Comment (niemeyer) [inline docs/interfaces.md]: True.. hmm.. how about `can-sandbox`?

**Comment by jdstrand** [comment]
Per IRC, rename as 'browser-support' and update the attribute to be 'allow-sandbox'

File: interfaces/builtin/browser.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+	switch securitySystem {
+	case interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityDBus, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *BrowserInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	allow_browser_sandbox := false
```
Comment (niemeyer) [inline interfaces/builtin/browser.go]: This will do all of the logic here:
```
allowSandbox, _ := plug.Attrs["allow-sandbox"].(bool)
```
Please note we use camelCase rather than under_lines.

File: interfaces/builtin/browser.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+func (iface *BrowserInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface.Name()))
+	}
+
+	// It's fine if allow-browser-sandbox isn't specified, but it it is,
+	// it needs to be bool
+	if _, ok := plug.Attrs["allow-browser-sandbox"]; ok {
```
Comment (niemeyer) [inline interfaces/builtin/browser.go]: Slightly simpler:
```
if v, ok := plug.Attrs["allow-sandbox"]; ok {
        if _, ok = v.(bool); !ok {
                return fmt.Errorf(...)
        }
}

File: interfaces/builtin/unity7.go (hunk 3 above)
Comment (jdstrand) [inline interfaces/builtin/unity7.go]: It is needed for Chrome to work so I included it here. I could do it as a separate PR if you prefer.
